watcher/src/watcher.rs

```diff
@@ -257,121 +267,126 @@ const MAX_BLOCKS_PER_SYNC_ITERATION: u32 = 1000;
     stop_requested: Arc<AtomicBool>,
+    join_handles: Vec<thread::JoinHandle<()>>,
+    logger: Logger,
 }
 
 impl WatcherSyncThread {
     /// Create a new watcher sync thread.
     pub fn new(
         watcher_db: WatcherDB,
-        ledger: impl Ledger + 'static,
+        ledger: LedgerDB,
         poll_interval: Duration,
         store_block_data: bool,
         logger: Logger,
     ) -> Result<Self, WatcherError> {
         log::debug!(logger, "Creating watcher sync thread.");
         let watcher = Watcher::new(watcher_db, store_block_data, logger.clone())?;
-
-        let currently_behind = Arc::new(AtomicBool::new(false));
+        let config_urls = watcher.watcher_db.get_config_urls()?;
         let stop_requested = Arc::new(AtomicBool::new(false));
-
-        let thread_currently_behind = currently_behind.clone();
-        let thread_stop_requested = stop_requested.clone();
-        let join_handle = Some(
-            thread::Builder::new()
-                .name("WatcherSync".into())
+        let ledger_ref = Arc::new(ledger);
+        let watcher_ref = Arc::new(watcher);
+        let mut join_handles = Vec::new();
+
+        for url in config_urls {
+            let thread_ledger_ref = ledger_ref.clone();
+            let thread_watcher_ref = watcher_ref.clone();
+            let thread_poll_interval = poll_interval.clone();
+            let thread_stop_requested = stop_requested.clone();
+            let thread_logger = logger.clone();
+
+            let join_handle = thread::Builder::new()
+                .name(url.to_string())
                 .spawn(move || {
-                    Self::thread_entrypoint(
-                        ledger,
-                        watcher,
-                        poll_interval,
-                        thread_currently_behind,
+                    Self::sync_watcher_node(
+                        url,
+                        thread_ledger_ref,
+                        thread_watcher_ref,
+                        thread_poll_interval,
                         thread_stop_requested,
-                        logger,
+                        thread_logger,
                     );
                 })
-                .expect("Failed spawning WatcherSync thread"),
-        );
+                .expect("Could not start watcher threads");
+
+            join_handles.push(join_handle)
+        }
 
         Ok(Self {
-            join_handle,
-            currently_behind,
             stop_requested,
+            join_handles,
+            logger,
         })
     }
 
     /// Stop the watcher sync thread.
     pub fn stop(&mut self) {
         self.stop_requested.store(true, Ordering::SeqCst);
-        if let Some(thread) = self.join_handle.take() {
-            thread.join().expect("WatcherSync thread join failed");
+        for handle in self.join_handles.drain(..) {
+            match handle.join() {
+                Ok(_) => {}
+                Err(err) => {
+                    log::debug!(
```

Comment:
The original behavior was to panic if joining fails, I don't think there's a legitimate reason for it to ever fail so I am leaning towards preserving that or maybe stopping and returning an error (so the caller can decide what to do about it).

Reply:
I would certainly agree to that and think that makes sense, I wrote this in so that if a single thread panicked it wouldn't orphan the rest of the thread joins, but I think we can just assume it won't panic